src/tree.ml

```diff
@@ -211,6 +201,28 @@ module CP = struct
     if res = None then choose_vf pools else res
 end
 
+(** {1 Variable referencing and display} *)
+
+(** Hashable term variables. *)
+module TvarHash =
+struct
+  type t = tvar
+  let equal = Bindlib.eq_vars
+  let hash = Bindlib.hash_var
+end
+
+(** Hashtable of term variables. *)
+module TvarHashtbl = Hashtbl.Make(TvarHash)
+
+(** Associates to each variable an identifier to be used by
+    {!constructor:Tree_types.TC.Vari}. *)
+let varmap : int TvarHashtbl.t = TvarHashtbl.create 0
+
+(** Prefix of bound variables. *)
+let var_prefix = "var"
+
+(** A counter for identifiers. TODO hide it in a [store] function. *)
+let vid : int Pervasives.ref = Pervasives.ref 0
```

Comment:
Is the hash table and the identifier ever reset? Otherwise you will always have growing number which will not be easily readable when printing.

With the hash table, you will also be leaking memory since the variables, even when not used anymore, will still be accessible (and hence not collected by the GC). To work around, you can use instead a weak hash table (`Weak` module). But a better solution would be to reset the hash table and the counter when you have finished building a decision tree (so that next time it starts fresh, from 0).

Reply:
But the ids might be accessed when evaluating the decision trees, in `eval.ml`, so we don't want to reset the hash table

Reply:
OK. Having one per tree would make more sense indeed.

Reply:
All this looks over complicated to me. @rlepigre what's wrong with Gabriel's first proposal? Note that all this is just to get clearer graphs. It's only for debug purposes.

Reply:
What is wrong is that it makes a bad use of Bindlib. I don't want the code to contain hacks like these because it might encourage others to do similar things (thinking that it is the right way to use Bindlib).

Reply:
I think it tends to be clearer to explicitly declare the identifier used to relate the variables than hiding this identifiers in their names (which can be seen as hackish).

Reply:
I understand but, in this case, is it worth doing that? it will increase memory and slow down tree construction, unless this mapping is done only when one wants to print the trees.

Reply:
Hash table operations are roughly constant time. And the tree construction is not efficiency critical anyway: only matching is.

Reply:
But this seems tricky since it makes the dtree type depend on a functor, doesn't it force to raise everything to the functor level?

Reply:
What you are missing here is that the custom hash table module is only created once (using the functor) at the toplevel of the file. But then you will just manipulate distinct hash tables (all created with `TvarHashtbl.create`).

Reply:
But if you want one hash table per tree, don't you want to modify the `dtree` type to contain the hash table?

Reply:
Sure, but it is only gonna include an element of type `int TvarHashtbl.t` somewhere. No need to apply the functor more than once (at the top level of the file).